src/lighthouse.rs

```diff
@@ -89,56 +112,94 @@ fn quorum_changed(a: &Vec<QuorumMember>, b: &Vec<QuorumMember>) -> bool {
     return a_ids != b_ids;
 }
 
-// Checks whether the quorum is valid and an explanation for the state.
-fn quorum_valid(state: &RoomState, opt: &LighthouseOpt) -> (bool, String) {
-    let mut first_joined = Instant::now();
+// Checks whether the quorum is valid, the new quorum and an explanation for the state.
+fn quorum_compute(
+    now: Instant,
+    heartbeats: &HashMap<String, Instant>,
+    state: &RoomState,
+    opt: &LighthouseOpt,
+) -> (Option<Vec<QuorumMember>>, String) {
+    let healthy_participants: HashMap<String, QuorumMemberDetails> = state
+        .participants
+        .clone()
+        .into_iter()
+        .filter(|(replica_id, _details)| {
+            let last_heartbeat = heartbeats.get(replica_id);
+            if last_heartbeat.is_none() {
+                return false;
+            }
 
-    for details in state.participants.values() {
-        if details.joined < first_joined {
-            first_joined = details.joined;
-        }
-    }
+            now.duration_since(*last_heartbeat.unwrap())
+                < Duration::from_millis(opt.heartbeat_timeout_ms)
+        })
+        .collect();
+
+    let mut candidate_participants: Vec<QuorumMember> = healthy_participants
+        .values()
+        .map(|details| details.member.clone())
+        .collect();
 
+    // Sort by replica ID to get a consistent ordering across runs.
+    candidate_participants.sort_by_key(|p| p.replica_id.clone());
+
+    let metadata = format!(
+        "[{}/{} participants healthy]",
+        healthy_participants.len(),
+        state.participants.len()
+    );
+
+    // Check if we can use the previous quorum.
     if state.prev_quorum.is_some() {
-        let mut is_fast_quorum = true;
         let prev_quorum = state.prev_quorum.as_ref().unwrap();
 
-        for prev_member in prev_quorum.participants.iter() {
-            if !state.participants.contains_key(&prev_member.replica_id) {
-                is_fast_quorum = false;
-                break;
-            }
-        }
+        // Fast quorum is when all previous participants are still in the quorum
+        // and we have enough participants to form a quorum.
+        let is_fast_quorum = prev_quorum
+            .participants
+            .iter()
+            .all(|prev_member| healthy_participants.contains_key(&prev_member.replica_id));
 
         if is_fast_quorum {
-            return (is_fast_quorum, format!("Fast quorum found!"));
+            return (
+                Some(candidate_participants),
+                format!("Fast quorum found! {}", metadata),
+            );
         }
     }
 
-    if state.participants.len() < opt.min_replicas as usize {
+    if healthy_participants.len() < opt.min_replicas as usize {
```

Comment:
Do we maintain a sequence number for each participant? For example, we could have a race where:

1) min members == 5
2) currently 4 have heartbeated
3) the fifth is just about to heartbeat, but then this code runs first and the fifth gets excluded
4) repeat over and over

If we have sequence numbers for each checkin, then if the fifth one is on a lower sequence number (but still healthy), we know we should keep waiting until the fifth one bumps the sequence number up to the expected sequence number.

Reply:
We don't have any sequence numbers -- this logic is structured off of Raft heartbeats which assumes that the heartbeating is much faster than the timeout interval. Default settings is to heartbeat every 100ms and unhealthy timeout at 1s.

We do have waiting logic but that only considers healthy replicas